azure-iot-device/tests/common/transport/test_pipeline_stages_base.py

```diff
@@ -0,0 +1,544 @@
+from azure.iot.device.common.transport import pipeline_ops_base
+from azure.iot.device.common.transport import pipeline_events_base
+
+logging.basicConfig(level=logging.INFO)
+
+
+def assert_default_attributes(obj):
+    assert obj.name is obj.__class__.__name__
+    assert obj.next is None
+    assert obj.previous is None
+    assert obj.pipeline_root is None
+
+
+# because PipelineStage is abstract, we need something concrete
+class PipelineStage(pipeline_stages_base.PipelineStage):
+    def _run_op(self, op):
+        pass
+
+
+def Op():
+    return pipeline_ops_base.PipelineOperation()
+
+
+def get_fake_error():
+    return BaseException()
+
+
+@pytest.fixture
+def stage(mocker):
+    def stage_run_op(self, op):
+        if getattr(op, "action", None) is None or op.action == "pass":
+            self.complete_op(op)
+        elif op.action == "fail":
+            raise get_fake_error()
+        elif op.action == "pend":
+            pass
+        else:
+            assert False
+
+    first_stage = PipelineStage()
+    first_stage.unhandled_error_handler = mocker.Mock()
+    first_stage._run_op = functools.partial(stage_run_op, first_stage)
+    mocker.spy(first_stage, "_run_op")
+    mocker.spy(first_stage, "run_op")
+
+    next_stage = PipelineStage()
+    next_stage._run_op = functools.partial(stage_run_op, next_stage)
+    mocker.spy(next_stage, "_run_op")
+    mocker.spy(next_stage, "run_op")
+
+    first_stage.next = next_stage
+    first_stage.pipeline_root = first_stage
+
+    next_stage.previous = first_stage
+    next_stage.pipeline_root = first_stage
+
+    return first_stage
+
+
+@pytest.fixture
+def callback(mocker):
+    return mocker.Mock()
+
+
+@pytest.fixture
+def op():
+    op = Op()
+    op.name = "op"
+    return op
+
+
+@pytest.fixture
+def op2():
+    op = Op()
+    op.name = "op2"
+    return op
+
+
+@pytest.fixture
+def op3():
+    op = Op()
+    op.name = "op3"
+    return op
+
+
+@pytest.fixture
+def finally_op():
+    op = Op()
+    op.name = "finally_op"
+    return op
+
+
+@pytest.fixture
+def fake_error():
+    return get_fake_error()
+
+
+@pytest.fixture
+def event():
+    ev = pipeline_events_base.PipelineEvent()
+    ev.name = "test event"
+    return ev
+
+
+def assert_callback_succeeded(callback, op):
+    callback.assert_called_once_with(op)
+    assert op.error is None
+
+
+def assert_callback_failed(callback, op, error=None):
+    callback.assert_called_once_with(op)
+    if error:
+        assert op.error is error
+    else:
+        assert op.error is not None
+
+
+@pytest.mark.describe("PipelineStage initializer")
+class TestPipelineStageInitializer(object):
+    @pytest.mark.it("Sets required and default arguments correctly")
+    def test_initializer(self):
+        obj = PipelineStage()
+        assert_default_attributes(obj)
+
+
+@pytest.mark.describe("PipelineStage RunOp function")
+class TestPipelineStageRunOp(object):
+    @pytest.mark.it("calls _run_op")
+    def test_1(self, stage, op):
```

Comment:
>test_1 [](start = 8, length = 6)

Even though we have the `pytest.mark.it` to tell us what the test actually is testing for, I think the testname should still be something that's legible, for a couple reasons:

1) numbering tests gets out of control quickly as test are added and removed.
2) I think there's value to the `pytest.mark.it` telling us what the requirement is, while the test name tells us what the test __does___ (since those are not exactly 1:1)